### DISCUSSION TITLE
Determining whether an element is solely a text element or not within the Ginja template?

### DISCUSSION THREAD

**Discussion by ethindp** [sissaschool/xmlschema#409] [Q&A] [answered]
I'm trying to construct a static code generator for C++ (because all the ones out there that I know of are either paid or they don't really support what I'm trying to do/don't have an intuitive API). However, I'm uncertain about a couple things:

* How do I determine if an element is solely a text element? I was thinking of checking if `len(xsd_element.attributes.values()) == 0 and len(xsd_element.iterchildren()) == 0`, but this seems like a really hacky way of doing things.
* How do I determine the C++ type? If language is cpp would it just be `cpp_type` as the filter?

**Comment by brunato** [comment, marked as answer]
Hi,

about the 1st question: you have to look at the type. Each type has a content type, that can be:

1. empty
2. simple
3. mixed
4. element-only

for checking if the content of the element is only a text use:

  `xsd_element.type.has_simple_content()` 

if you don't mind about empty cases add also `is_empty()`.

For checking if the element has declared attributes use `len(xsd_element.attributes) == 0` (or `not xsd_element.attributes`).

For the second question, if you define a subclass for C++ starting with:

```python
class CPPGenerator(AbstractGenerator):
    """A C++ code generator for XSD schemas."""

    formal_language = 'CPP'

    ...
```

the auto-generated filter will be `cpp_type`.

**Comment by ethindp** [comment]
Thank you! :)
